diff -urN linux-3.13/arch/x86/syscalls/syscall_64.tbl linux-3.13-assignment1/arch/x86/syscalls/syscall_64.tbl
--- linux-3.13/arch/x86/syscalls/syscall_64.tbl	2014-01-19 18:40:07.000000000 -0800
+++ linux-3.13-assignment1/arch/x86/syscalls/syscall_64.tbl	2018-02-10 03:00:43.168013443 -0800
@@ -320,6 +320,8 @@
 311	64	process_vm_writev	sys_process_vm_writev
 312	common	kcmp			sys_kcmp
 313	common	finit_module		sys_finit_module
+314	64	hello			sys_hello
+315	64	sh_task_info		sys_sh_task_info
 
 #
 # x32-specific system call numbers start at 512 to avoid cache impact
diff -urN linux-3.13/hello/hello.c linux-3.13-assignment1/hello/hello.c
--- linux-3.13/hello/hello.c	1969-12-31 16:00:00.000000000 -0800
+++ linux-3.13-assignment1/hello/hello.c	2018-02-08 04:53:34.451940177 -0800
@@ -0,0 +1,6 @@
+#include <linux/kernel.h>
+
+asmlinkage long sys_hello(void) {
+	printk("Hello world\n");
+	return 0;
+}
diff -urN linux-3.13/hello/Makefile linux-3.13-assignment1/hello/Makefile
--- linux-3.13/hello/Makefile	1969-12-31 16:00:00.000000000 -0800
+++ linux-3.13-assignment1/hello/Makefile	2018-02-08 04:54:50.831938732 -0800
@@ -0,0 +1 @@
+obj-y := hello.o
diff -urN linux-3.13/include/linux/syscalls.h linux-3.13-assignment1/include/linux/syscalls.h
--- linux-3.13/include/linux/syscalls.h	2014-01-19 18:40:07.000000000 -0800
+++ linux-3.13-assignment1/include/linux/syscalls.h	2018-02-10 06:35:44.271096012 -0800
@@ -847,4 +847,6 @@
 asmlinkage long sys_kcmp(pid_t pid1, pid_t pid2, int type,
 			 unsigned long idx1, unsigned long idx2);
 asmlinkage long sys_finit_module(int fd, const char __user *uargs, int flags);
+asmlinkage long sys_hello(void);
+asmlinkage long sys_sh_task_info(long pid, char *filename);
 #endif
diff -urN linux-3.13/kernel/sys.c linux-3.13-assignment1/kernel/sys.c
--- linux-3.13/kernel/sys.c	2014-01-19 18:40:07.000000000 -0800
+++ linux-3.13-assignment1/kernel/sys.c	2018-02-11 01:04:38.479417909 -0800
@@ -2160,3 +2160,62 @@
 	return 0;
 }
 #endif /* CONFIG_COMPAT */
+
+// Referred: "ytliu.info/notes/linux/file_ops_in_kernel.html" for file i/o.
+
+SYSCALL_DEFINE2(sh_task_info, long, pid, char*, filename) {
+	struct task_struct *proc;
+	struct tty_struct *tty;
+	int flag = 0;
+	long errno = 0;
+
+	for_each_process(proc) {
+		if (proc != NULL) {
+			long tempPID = (long) task_pid_nr(proc);
+
+			if (tempPID == pid) {
+				char str[100];
+				mm_segment_t oldFS;
+				struct file *f = NULL;
+				int fd;
+				loff_t offset = 0;
+
+				flag = 1;
+
+				sprintf(str, "Process: %s\r\nPID: %ld\r\nProcess state: %ld\r\nPriority: %ld\r\n", proc -> comm, tempPID, (long) proc -> state, (long) proc -> prio);
+
+				oldFS = get_fs();
+				set_fs(KERNEL_DS);
+				fd = sys_open(filename, O_WRONLY, 0666);
+
+				if (fd >= 0) {
+					sys_write(fd, str, strlen(str));
+					f = fget(fd);
+					vfs_write(f, str, strlen(str), &offset);
+					fput(f);
+					sys_close(fd);
+				}
+				else {
+					errno = -ENOENT;
+					break;
+				}
+				set_fs(oldFS);
+
+				tty = get_current_tty();
+				if (tty != NULL) {
+					(tty -> driver -> ops -> write)(tty, str, strlen(str));
+				}
+				else {
+					errno = -ENOTTY;
+				}
+			}
+		}
+	}
+
+	if (flag == 0) {
+		errno = -ESRCH;
+	}
+
+	printk("errno: %ld\n", errno);
+	return errno;
+}
diff -urN linux-3.13/Makefile linux-3.13-assignment1/Makefile
--- linux-3.13/Makefile	2014-01-19 18:40:07.000000000 -0800
+++ linux-3.13-assignment1/Makefile	2018-02-08 04:56:19.623937051 -0800
@@ -754,8 +754,8 @@
 
 
 ifeq ($(KBUILD_EXTMOD),)
-core-y		+= kernel/ mm/ fs/ ipc/ security/ crypto/ block/
-
+core-y		+= kernel/ mm/ fs/ ipc/ security/ crypto/ block/ hello/
+  
 vmlinux-dirs	:= $(patsubst %/,%,$(filter %/, $(init-y) $(init-m) \
 		     $(core-y) $(core-m) $(drivers-y) $(drivers-m) \
 		     $(net-y) $(net-m) $(libs-y) $(libs-m)))
diff -urN linux-3.13/security/tomoyo/builtin-policy.h linux-3.13-assignment1/security/tomoyo/builtin-policy.h
--- linux-3.13/security/tomoyo/builtin-policy.h	1969-12-31 16:00:00.000000000 -0800
+++ linux-3.13-assignment1/security/tomoyo/builtin-policy.h	2018-02-08 05:24:38.467904894 -0800
@@ -0,0 +1,12 @@
+static char tomoyo_builtin_profile[] __initdata =
+"";
+static char tomoyo_builtin_exception_policy[] __initdata =
+"initialize_domain /sbin/modprobe from any\n"
+"initialize_domain /sbin/hotplug from any\n"
+"";
+static char tomoyo_builtin_domain_policy[] __initdata =
+"";
+static char tomoyo_builtin_manager[] __initdata =
+"";
+static char tomoyo_builtin_stat[] __initdata =
+"";
diff -urN linux-3.13/security/tomoyo/policy/exception_policy.conf linux-3.13-assignment1/security/tomoyo/policy/exception_policy.conf
--- linux-3.13/security/tomoyo/policy/exception_policy.conf	1969-12-31 16:00:00.000000000 -0800
+++ linux-3.13-assignment1/security/tomoyo/policy/exception_policy.conf	2018-02-08 05:24:18.091905279 -0800
@@ -0,0 +1,2 @@
+initialize_domain /sbin/modprobe from any
+initialize_domain /sbin/hotplug from any
